pybamm/util.py

```diff
@@ -345,3 +345,25 @@ def install_jax(arguments=None):  # pragma: no cover
             f"jaxlib>={JAXLIB_VERSION}",
         ]
     )
+
+# https://docs.pybamm.org/en/latest/source/user_guide/contributing.html#managing-optional-dependencies-and-their-imports
+def have_optional_dependency(module_name, attribute=None):
+    try:
+        # Attempt to import the specified module
+        module = importlib.import_module(module_name)
+
+        if attribute:
+            # If an attribute is specified, check if it's available
+            if hasattr(module, attribute):
+                imported_attribute = getattr(module, attribute)
+                return imported_attribute  # Return the imported attribute
+            else:
+                # Raise an ModuleNotFoundError if the attribute is not available
+                raise ModuleNotFoundError(f"Optional dependency {module_name} is not available. See https://docs.pybamm.org/en/latest/source/user_guide/installation/index.html#optional-dependencies for more details.")
```

Comment:
The message string should be refactored outside the `try-except` block, something like this -
```py
def have_optional_dependency(module_name, attribute=None):
  err_msg = f"Optional dependency {module_name} ...."
  try:
  ...
  raise ModuleNotFoundError(err_msg)
  ...
  raise ModuleNotFoundError(err_msg)
```

This -
1. Avoids repetition in code
2. For longer error messages, the message string itself should not be passed to the error class, rather a variable should be passed in for better readability.
  For example -
    ```py
    sub = "Some value"
    raise RuntimeError(f"{sub!r} is incorrect")
    ```
    prints
    ```py
    Traceback (most recent call last):
      File "tmp.py", line 2, in <module>
        raise RuntimeError(f"{sub!r} is incorrect")
    RuntimeError: 'Some value' is incorrect
    ```
    See how the error message is repeated twice. The readability gets worse as we increase the size of the string.

Reply:
Done, Got it 🚀